zilencer/management/commands/calculate_first_visible_message_id.py

```diff
@@ -0,0 +1,35 @@
+from typing import Any
+
+from django.core.management.base import CommandParser
+
+from zerver.lib.management import ZulipBaseCommand
+from zerver.lib.cache import realm_first_visible_message_id_cache_key, \
+    cache_get, cache_set
+from zerver.lib.message import estimiate_recent_messages
+from zerver.models import Realm, Message
+
+
+class Command(ZulipBaseCommand):
+    help = """Calculate the value of first visible message ID and store it in cache"""
+
+    def add_arguments(self, parser: CommandParser) -> None:
+        self.add_realm_args(parser)
+
+    def handle(self, *args: Any, **options: str) -> None:
+        realm = self.get_realm(options)
+
+        if realm is None:
+            realms = Realm.objects.all()
+        else:
+            realms = [realm]
+        for realm in realms:
+            cache_expired = cache_get(realm_first_visible_message_id_cache_key(realm)) is None
+            if realm.message_visibility_limit is not None and \
+                    (estimiate_recent_messages(realm, 1) > 0 or cache_expired):
+                try:
+                    first_visible_message_id = Message.objects.filter(sender__realm=realm).values('id').\
+                        order_by('-id')[realm.message_visibility_limit - 1]["id"]
+                except IndexError:
+                    first_visible_message_id = 0
+                cache_set(realm_first_visible_message_id_cache_key(realm),
+                          first_visible_message_id, timeout=3600*24*7)
```

Comment:
Let's set no timeout. The old value is always still good until more messages are sent, and even then it's more accurate than no value (which we interpret as 0.)